miniapps/dpg/convection-diffusion.cpp

```diff
@@ -0,0 +1,572 @@
+   a->AddTestIntegrator(new DiffusionIntegrator(bbtcoeff),
+                        TestSpace::v_space, TestSpace::v_space);
+   // c2 (τ,δτ)
+   a->AddTestIntegrator(new VectorFEMassIntegrator(c2_coeff),
+                        TestSpace::tau_space, TestSpace::tau_space);
+   // (∇⋅τ,∇⋅δτ)
+   a->AddTestIntegrator(new DivDivIntegrator(one),
+                        TestSpace::tau_space, TestSpace::tau_space);
+
+   FunctionCoefficient f(f_exact);
+   a->AddDomainLFIntegrator(new DomainLFIntegrator(f),TestSpace::v_space);
+
+   FunctionCoefficient hatuex(exact_hatu);
+   VectorFunctionCoefficient hatfex(dim,exact_hatf);
+   Array<int> elements_to_refine;
+   FunctionCoefficient uex(exact_u);
+   VectorFunctionCoefficient sigmaex(dim,exact_sigma);
+
+   GridFunction hatu_gf, hatf_gf;
+
+   socketstream u_out;
+   socketstream sigma_out;
+
+   double res0 = 0.;
+   double err0 = 0.;
+   int dof0;
+   std::cout << "\n  Ref |"
+             << "    Dofs    |"
+             << "  L2 Error  |"
+             << "  Rate  |"
+             << "  Residual  |"
+             << "  Rate  |" << endl;
+   std::cout << std::string(64,'-') << endl;
+
+   if (static_cond) { a->EnableStaticCondensation(); }
+   for (int it = 0; it<=ref; it++)
+   {
+      a->Assemble();
+
+      Array<int> ess_tdof_list_uhat;
+      Array<int> ess_tdof_list_fhat;
+      Array<int> ess_bdr_uhat;
+      Array<int> ess_bdr_fhat;
+      if (mesh.bdr_attributes.Size())
+      {
+         ess_bdr_uhat.SetSize(mesh.bdr_attributes.Max());
+         ess_bdr_fhat.SetSize(mesh.bdr_attributes.Max());
+         ess_bdr_uhat = 1; ess_bdr_fhat = 0;
+         if (prob == prob_type::EJ)
+         {
+            ess_bdr_uhat = 0;
+            ess_bdr_fhat = 1;
+            ess_bdr_uhat[1] = 1;
+            ess_bdr_fhat[1] = 0;
+         }
+         hatu_fes->GetEssentialTrueDofs(ess_bdr_uhat, ess_tdof_list_uhat);
+         hatf_fes->GetEssentialTrueDofs(ess_bdr_fhat, ess_tdof_list_fhat);
+      }
+
+      // shift the ess_tdofs
+      int n = ess_tdof_list_uhat.Size();
+      int m = ess_tdof_list_fhat.Size();
+      Array<int> ess_tdof_list(n+m);
+      for (int j = 0; j < n; j++)
+      {
+         ess_tdof_list[j] = ess_tdof_list_uhat[j]
+                            + u_fes->GetTrueVSize()
+                            + sigma_fes->GetTrueVSize();
+      }
+      for (int j = 0; j < m; j++)
+      {
+         ess_tdof_list[j+n] = ess_tdof_list_fhat[j]
+                              + u_fes->GetTrueVSize()
+                              + sigma_fes->GetTrueVSize()
+                              + hatu_fes->GetTrueVSize();
+      }
+
+      Array<int> offsets(5);
+      offsets[0] = 0;
+      int dofs = 0;
+      for (int i = 0; i<trial_fes.Size(); i++)
+      {
+         offsets[i+1] = trial_fes[i]->GetVSize();
+         dofs += trial_fes[i]->GetTrueVSize();
+      }
+      offsets.PartialSum();
+
+      BlockVector x(offsets); x = 0.0;
+      hatu_gf.MakeRef(hatu_fes,x.GetBlock(2));
```

Comment:
There are many compiler warnings about the deprecated conversion to `double *` in this and other miniapp files, which can be fixed by using `x.GetBlock(2).GetData()`.

Reply:
One can also use this version of `MakeRef` (which should work on device too):
```c++
   virtual void MakeRef(FiniteElementSpace *f, Vector &v, int v_offset);
```

Reply:
Fixed in [1b49257](https://github.com/mfem/mfem/pull/3234/commits/1b492576e13804c9f14cea5c5b4d36c8178c1c33)